gr-picoscope/picoscope.hpp

```diff
@@ -0,0 +1,559 @@
+    bool               auto_arm                 = true;
+    ChannelMap         enabled_channels;
+    PortMap            enabled_ports;
+    trigger_setting_t  trigger;
+};
+
+struct State {
+    std::vector<Channel>        channels;
+    std::atomic<std::size_t>    data_available     = 0;
+    std::atomic<bool>           data_finished      = false;
+    bool                        initialized        = false;
+    bool                        configured         = false;
+    bool                        closed             = false;
+    bool                        armed              = false;
+    bool                        started            = false; // TODO transitional until nodes have state handling
+    int16_t                     handle             = -1;    ///< picoscope handle
+    int16_t                     overflow           = 0;     ///< status returned from getValues
+    int16_t                     max_value          = 0;     ///< maximum ADC count used for ADC conversion
+    double                      actual_sample_rate = 0;
+    std::thread                 poller;
+    BufferHelper<Error>         errors;
+    std::atomic<poller_state_t> poller_state = poller_state_t::IDLE;
+    std::atomic<bool>           forced_quit  = false; // TODO transitional until we found out what
+                                                      // goes wrong with multithreaded scheduler
+    std::size_t produced_worker = 0;                  // poller/callback thread
+};
+
+} // namespace detail
+
+// optional shortening
+template<typename T, fair::meta::fixed_string description = "", typename... Arguments>
+using A = fair::graph::Annotated<T, description, Arguments...>;
+
+using fair::graph::Visible;
+
+template<typename PSImpl>
+struct Picoscope : public fair::graph::node<PSImpl, fair::graph::BlockingIO<true>> {
+    A<std::string, "serial number", Visible> serial_number;
+    A<double, "sample rate", Visible>        sample_rate = 10000.;
+    // TODO any way to get custom enums into pmtv??
+    A<std::string, "acquisition mode as string", Visible>      acquisition_mode_string  = std::string("STREAMING");
+    A<std::size_t, "pre-samples", Visible>                     pre_samples              = 1000;
+    A<std::size_t, "post-samples", Visible>                    post_samples             = 9000;
+    A<std::size_t, "no. captures (rapid block mode)", Visible> rapid_block_nr_captures  = 1;
+    A<bool, "trigger once (rapid block mode)", Visible>        trigger_once             = false;
+    A<double, "poll rate (streaming mode)", Visible>           streaming_mode_poll_rate = 0.001;
+    A<bool, "auto-arm", Visible>                               auto_arm                 = true;
+    A<ChannelMap, "enabled analog channels", Visible>          enabled_channels;
+    A<PortMap, "enabled digital ports", Visible>               enabled_ports;
+    A<trigger_setting_t, "trigger settings", Visible>          trigger;
```

Comment:
trigger_setting_t is not a pmt_t compatible type -> must be broken up into it's constituents.

similar for most of the other settings.

Reply:
this one could be split, but it get's nastier for enabled_channels and enabled_ports, which are maps of structs (could also be vector of structs, same problem).
Should I use property_map of property_maps (or vector of maps) for those, using strings instead of enums? the enum problem I also have with acquisition_mode, which is currently exposed as string.

Reply:
What's wrong with making `enabled_channels` a `std::vector<bool>` ... would be simple and IMO intuitive.

Reply:
it now uses separate std::vector<std::string>, std::vector<double> (one entry for each channel) for the various channel properties.